stdlib/public/core/StringCharacterView.swift

```diff
@@ -276,39 +194,118 @@ extension String.CharacterView : BidirectionalCollection {
   /// 
   /// In an empty character view, `startIndex` is equal to `endIndex`.
   public var startIndex: Index {
-    return Index(_base: unicodeScalars.startIndex)
+    return Index(_base: unicodeScalars.startIndex, in: self)
   }
 
   /// A character view's "past the end" position---that is, the position one
   /// greater than the last valid subscript argument.
   ///
   /// In an empty character view, `endIndex` is equal to `startIndex`.
   public var endIndex: Index {
-    return Index(_base: unicodeScalars.endIndex)
+    return Index(_base: unicodeScalars.endIndex, in: self)
   }
 
   /// Returns the next consecutive position after `i`.
   ///
   /// - Precondition: The next position is valid.
   public func index(after i: Index) -> Index {
-    _precondition(i._base != i._base._viewEndIndex, "cannot increment endIndex")
-    return Index(_base: i._endBase)
+    _precondition(i._base != unicodeScalars.endIndex,
```

Comment:
Since the index does not have the reference to the underlying storage, this check is not sufficient.  The user can be taking an index from one sting and passing it to other's `index(after:)` method.  This method should perform a full bounds check.